src/console/man/om_sampler.1
.\" Text automatically generated by txt2man
.TH om_sampler  "02 April 2013" "" ""
.SH NAME
\fBom_sampler \fP- dump samples from a specified source
\fB
.SH SYNOPSIS
.nf
.fam C
     \fBom_sampler\fP [-] \fIv\fP \fB--version\fP | \fIs\fP \fB--source\fP \fIFILE\fP ] [ \fB--dump-env\fP \fB--cell-start\fP \fINUM\fP \fB--cell-end\fP \fINUM\fP ] [ \fB--log-level\fP \fILEVEL\fP ]

.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
\fBom_sampler\fP is a command line tool to dump openModeller sample data given a source file specifying sampler parameters. A sample is a vector of environmental data associated with a point in the geographical space. Source files must include references to locality points and environmental layers, which constitute the basic parameters to the openModeller sampler. Accepted formats for source files are request files following the same pattern used to run om_console or XML files containing serialized models or model request parameters.
.PP
By default, \fBom_sampler\fP will dump all data associated with the presence and/or absence points specified in the source file. However, if the "dump-env" parameter is present, then all environment samples will be dumped for each cell of the mask file. Start and end positions can be specified in this case. The first cell (position zero) is the top left most. Iteration happens from left to right, and from top to bottom in the raster. When dumping the environment, the coordinates displayed correspond to the centroid of the cell. When "nodata" is displayed for a given sample, this means that at least one of the layers (including the mask) has no value in the location.
.SH OPTIONS
.TP
.B
-\fIv\fP, \fB--version\fP
Display version info.
.TP
.B
-\fIs\fP, \fB--source\fP
Source file with sampler parameters (locality data and layers). It must be a path in the file system pointing to a request file following the same pattern used to run om_console or an XML file containing serialized models or model request parameters.
.TP
.B
\fB-e\fP, \fB--dump-env\fP
Dump environment data for all cells of the mask file instead of dumping locality data.
.TP
.B
\fB--cell-start\fP
Start at cell (default 0). Used only with dump-env.
.TP
.B
\fB--cell-end\fP
End at cell (default 1000). Used only with dump-env.
.TP
.B
\fB--log-level\fP
openModeller log level: debug, warn, info or error. Defaults to "info".
.SH AUTHORS
Renato De Giovanni <renato at cria dot org dot br>


ChangeLog
Release 1.4.0 (?-?-?)
-------------
SVN revision: ?

	* New CONSENSUS algorithm that can be used to generate models with different algorithms by converting each individual model into a binary model given a threshold and then "merging" them into a final combined model according to a specified level of agreement.
	* Upgraded gSOAP version being used by OMWS to 2.8.13.
	* Included library API versioning, starting with 0 from this release.
	* Moved some of the compiler definitions from CMake to internal #defines, so that client programs can include openModeller headers without worrying about adding specific definitions if they want to compile directly without using CMake.
	* New command-line tool om_evaluate to return raw model values given a set of points and an environmental scenario.
	* Fixed segfault when needNormalization was called on a deserialized SVM algorithm.
	* Modified om_pseudo command-line tool to also accept XML input (in this case also producing XML output), optionally storing it into a file. When a model is specified as a parameter, presences are generated inside the model, while absences are generated outside.

Release 1.3.0 (2012-07-23)
-------------
SVN revision: 5592

	* New version of OM-MAXENT compatible with Maxent 3.3.3e.
	* Changed the deserialization code to accept XML with namespace prefixes (they are now ignored if present).
	* Fixed bug in ANN algorithm which was limiting the number of input points to 10k (seg faulting otherwise).
	* New documentation about the web service interface (OMWS), aka SOAP interface.
	* New documentation about oM Server (server-side implementation of OMWS).
	* Fixed bug in oM Server getProgress operation which was sometimes returning 100% when the job was still unfinished.
	* New oM Server configuration/script allowing automatic synchronization of layers.
	* Several improvements in the OMWS Perl command-line client.
	* Several adjustments in the SWIG Python interface, exposing more methods and objects.
	* Fixed memory leak in the Environmental Distance algorithm when the Mahalanobis metric was used after model deserialization.
	* New WCS proxy raster driver that allows WCS rasters to be identified through strings in the format: wcs>wcs_service_url>coverage_id

Release 1.2.0 (2011-07-21)
--------------------------
SVN revision: 5330

	* Changed the ENFA algorithm to force marginality to be the first factor.  Changed broken stick distribution to work on specialisation factors only.  Add new parameter to enable log display of variable loadings for retained factors.
	* om_console now accepts a second parameter (after the request file) to specify the log level.
	* Changed the ENFA algorithm allowing absence data to be treated as background points, and log transforming the geomean calculation to avoid the occasional float overflow with large numbers of presences. Fixed a bug in the RetainVariation parameter (range changed from 50-100 to 0.5-1.0). Also fixed a deserialization error.
	* Compilation now requires CMake 2.6.4.
	* New algorithm Random Forests.
	* Web service interface now accepts a new optional parameter specifying the prediction threshold to be used when calculating area statistics during projection.
	* Renamed envelope element in the getProjectionMetadata response of the web service API to distinguish it from the subelement.
	* Added control to a maximum number of concurrent jobs on the modelling server.
	* Environmental values can now be loaded from the Sampler XML when provided.
	* Web service interface now accepts optional parameters: IgnoreAbsences for the confusion matrix and UseAbsencesAsBackground for the ROC curve calculations.
	* Added support to gzip compression in the Web service interface.
	* Changed the Environmental Distance algorithm to return chi-square distribution probabilities when the metric is Mahalanobis and the maximum distance is 1.
	* Changed the Maxent algorithm allowing absence data to be treated as background points. Also fixed memory deallocation issues.

Release 1.1.0 (2010-02-05)
--------------------------
SVN revision: 5114

	* Bugfix: Interquartile and envelope size adjustments were using wrong reference values for salinity and temperature in the Aquamaps algorithm.
	* Bugfix: All GARP algorithms were crashing when the abundance value associated with one of the occurrence points had more than two digits in the integer part.
	* Increased the default number of neurons in the hidden layer of the ANN algorithm.
	* New argument "s" in om_niche allowing model outputs to be scaled.
	* Re-implemented and re-enabled Maximum entropy algorithm based on the well-known Maxent.
	* Changed environmental distance algorithm forcing the Mahalanobis distance to use an identity matrix as the covariance matrix when there's only one input point.
	* Fixed small logical error in the CSM algorithm that could affect models with a large number of layers.
	* New algorithm (Niche Mosaic) implemented by Dr. Missae Yamamoto.
	* New algorithm (ENFA) kindly implemented by Dr. Chris Yesson.

Release 1.0 (2009-05-22)
------------------------
SVN revision: 5035

	* Changed Artificial Neural Networks algorithm to avoid exceeding a maximum number of iterations when training by minimum error (parameter Epoch is used to specify the maximum number of iterations).
	* Added new method in the ConfusionMatrix class to use the lowest probability among the training presences as the threshold.
	* Changed the modelling protocol to allow createModel and testModel requests to specify what statistics should be calculated.
	* Removed model statistics parameters from om_model since they can be specified as part of the XML request.
	* All command line programs now return the same openModeller version number.
	* Changed om_niche parameters (now using getopts) and created a man page for it.
	* Implemented custom log file in the modelling server.

src/console/om_sampler.cpp
#include <openmodeller/om.hh>
#include <openmodeller/Configuration.hh>
#include <openmodeller/os_specific.hh>
#include <openmodeller/file_parser.hh>

#include "request_file.hh"

#include "getopts/getopts.h"

#include "om_cmd_utils.hh"

#include <istream>
#include <stdlib.h>
#include <string>
#include <stdio.h>
#include <iostream>  // I/O 
#include <fstream>   // file I/O

#include <stdexcept>
 
using namespace std;

void printOccurrences( ostream & stream, ConstOccurrencesPtr occ );

int main( int argc, char **argv ) {

  Options opts;
  int option;

  // command-line parameters (short name, long name, description, take args)
  opts.addOption( "v", "version"   , "Display version info"                        , false );
  opts.addOption( "s", "source"    , "Source with references to points and layers" , true );
  opts.addOption( "e", "dump-env"  , "Dump environment data for a specified cell range", false );
  opts.addOption( "", "cell-start" , "Cell position to start environment dumping (default 0)"  , true );
  opts.addOption( "", "cell-end"   , "Cell position to end environment dumping (default 1000)" , true );
  opts.addOption( "", "log-level"  , "Set the log level (debug, warn, info, error)", true );

  std::string log_level("info");
  std::string source("");
  bool        dump_env = false;
  std::string start_string("0");
  std::string end_string("1000");

  if ( ! opts.parse( argc, argv ) ) {

    opts.showHelp( argv[0] ); 
    exit(0);
  }

  // Set up any related external resources
  setupExternalResources();

  OpenModeller om;

  while ( ( option = opts.cycle() ) >= 0 ) {

    switch ( option ) {

      case 0:
        printf( "om_sampler %s\n", om.getVersion().c_str() );
        printf("This is free software; see the source for copying conditions. There is NO\n");
        printf("warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n");
        exit(0);
        break;
      case 1:
        source = opts.getArgs( option );
        break;
      case 2:
        dump_env = true;
        break;
      case 3:
        start_string = opts.getArgs( option );
        break;
      case 4:
        end_string = opts.getArgs( option );
        break;
      case 5:
        log_level = opts.getArgs( option );
        break;
      default:
        break;
    }
  }

  // Check requirements
  if ( source.empty() ) {

    printf( "Please specify a source (request file in txt or xml) with references to points and layers.\n");
    exit(-1);
  }

  int cell_start = 0;
  int cell_end = 0;

  if ( dump_env ) {

    cell_start = atoi( start_string.c_str() );

    cell_end = atoi( end_string.c_str() );
  }

  // Log stuff

  Log::Level level_code = getLogLevel( log_level );

  Log::instance()->setLevel( level_code );

  // Real work

  try {

    // Read source

    bool is_xml = false;

    // Try to open file and check if first char is "<". In this case we will
    // assume that it's an XML file.
    std::ifstream fs( source.c_str(), std::ios_base::in );

    if ( fs.fail() ) {

      printf( "Could not open source file.\n");
      exit(-1);
    }

    std::string line("");

    while ( getline( fs, line ) ) {

      for ( unsigned int i = 0; i < line.size(); ++i ) {    

        // Skip carriage returns, line feeds and spaces
        if ( line[i] == '\r' || line[i] == '\n' || line[i] == ' ' ) {

          continue;
        }

        // Looks like an XML file
        if ( line[i] == '<' ) {

          is_xml = true;
        }

        break;
      }
    }

    fs.close();

    SamplerPtr sampler;

    if ( is_xml ) {

      ConfigurationPtr config = Configuration::readXml( source.c_str() );

      sampler = createSampler( config->getSubsection( "Sampler" ) );
    }
    else {

      // Must be a txt request file, so parse it

      FileParser fp( source.c_str() );

      // Load points

      // Obtain the Well Known Text string for the localities
      // coordinate system.
      std::string oc_cs = fp.get( "WKT coord system" );

      // Get the name of the file containing localities
      std::string oc_file = fp.get( "Occurrences source" );

      // Get the label
      std::string oc_name = fp.get( "Occurrences group" );

      // When a model needs to be created, 'WKT coord system' and 
      // 'Species file' are mandatory parameters
      if ( oc_cs.empty() ) {

        printf( "'WKT coord system' keyword not specified in the request file!\n" );
        exit(-1);
      }

      if ( oc_file.empty() ) {

        printf( "'Occurrences source' keyword not specified in the request file!\n" );
        exit(-1);
      }

      // Populate the occurences list from the localities file
      OccurrencesReader* oc_reader = OccurrencesFactory::instance().create( oc_file.c_str(), oc_cs.c_str() );

      OccurrencesPtr presences = oc_reader->getPresences( oc_name.c_str() );

      OccurrencesPtr absences = oc_reader->getAbsences( oc_name.c_str() );

      delete oc_reader;

      // Load layers

      // Mask to select the desired species occurrence points
      std::string input_mask = fp.get( "Mask" );

      // Initiate the environment with all maps.
      std::vector<std::string> categ_map = fp.getAll( "Categorical map" );
      std::vector<std::string> cont_map = fp.getAll( "Map" );

      // When a model needs to be created, there should be at least one input map
      if ( ! (categ_map.size() + cont_map.size()) ) {

        printf( "At least one 'Map' or 'Categorical map' needs to be specified in the request file!\n" );
        exit(-1);
      }

      // If Mask was not specified, use the first layer
      if ( input_mask.empty() ) {

        if ( cont_map.size() ) {

          input_mask = cont_map[0];
        }
        else {

          input_mask = categ_map[0];
        }
      }

      EnvironmentPtr env = createEnvironment( categ_map, cont_map, input_mask );

      sampler = createSampler( env, presences, absences );

      std::string spatially_unique = fp.get( "Spatially unique" );
      if ( spatially_unique == "true" ) {

        sampler->spatiallyUnique();
      }

      std::string environmentally_unique = fp.get( "Environmentally unique" );
      if ( environmentally_unique == "true" ) {

        sampler->environmentallyUnique();
      }
    }

    // Print output

    int dim = sampler->numIndependent();

    // Header
    cout << "#id\tlabel\tlongitude\tlatitude\tabundance";

    for ( int i = 0; i < dim; ++i ) {

      cout << "\tattr" << i+1;
    }

    if ( dump_env ) {

      // Dump environment data

      EnvironmentPtr e = sampler->getEnvironment();

      // Initialize the iterator
      MapIterator it = e->getMask()->begin();

      // Initialize the terminal
      MapIterator fin;

      int cnt = 0;

      while( it != fin ) {

        if ( cnt < cell_start) {

          ++it;
          ++cnt;
          continue;
        }

        // Get the lon/lat coordinates from the iterator
        pair<Coord,Coord> lonlat = *it;

        // Extract the environment sample at that point
        Sample s = e->get( lonlat.first, lonlat.second );

        cout << "\n" << cnt << "\t" << "env data" << "\t" << lonlat.first << "\t" << lonlat.second << "\t" << "-";

        // Note:  s will have size() == 0 if the environment
        // does not have data for that location. This can happen
        // for a couple of reasons.
        //   1.  The location is not in the mask.
        //   2.  One of the layers has value == NoVal at that location.
        if ( s.size() > 0 ) {

          for ( int i = 0; i < dim; ++i ) {

            cout << "\t" << s[i];
          }
        }
        else {

            cout << "\t" << "nodata";
	}

        ++cnt;
        ++it;

        if ( cnt > cell_end ) {

          break;
        }
      }

      return 0;
    }

    // Default action: dump locality samples

    ConstOccurrencesPtr p = sampler->getPresences();

    if ( p ) {
    
      printOccurrences( cout, p );
    }

    ConstOccurrencesPtr a = sampler->getAbsences();

    if ( a ) {

      printOccurrences( cout, a );
    }

    return 0;
  }
  catch ( runtime_error e ) {

    printf( "om_sampler: %s\n", e.what() );
    exit(-1);
  }
}

// Print occurrences
void printOccurrences( ostream & stream, ConstOccurrencesPtr occ ) {

  int dim = occ->dimension();

  OccurrencesImpl::const_iterator it = occ->begin();
  OccurrencesImpl::const_iterator fin = occ->end();

  while ( it != fin ) {

    stream << "\n" << ((*it)->id()).c_str() << "\t" << occ->label() << "\t" << (*it)->x() << "\t" << (*it)->y() << "\t" << (*it)->abundance();

    Sample s = (*it)->environment();

    for ( int i = 0; i < dim; ++i ) {

      stream << "\t" << s[i];
    }

    it++;
  }
}
